File: leetcode/105.construct-binary-tree-from-preorder-and-inorder-traversal.ts
// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  const valInorderIdx = new Map(inorder.map((v, i) => [v, i]));

  const makeTree = (
    preLeftIdx: number,
    preRightIdx: number,
    inLeftIdx: number,
    inRightIdx: number,
  ): TreeNode | null => {
    if (preLeftIdx > preRightIdx) {
      return null;
    }

    const inRootIdx = valInorderIdx.get(preorder[preLeftIdx])!;
    const leftTreeSize = inRootIdx - inLeftIdx;

    return new TreeNode(
      preorder[preLeftIdx],
      makeTree(
        preLeftIdx + 1,
        preLeftIdx + leftTreeSize,
        inLeftIdx,
        inRootIdx - 1,
      ),
      makeTree(
        preLeftIdx + leftTreeSize + 1,
        preRightIdx,
        inRootIdx + 1,
        inRightIdx,
      ),
    );
  };

  return makeTree(0, preorder.length - 1, 0, inorder.length - 1);
}
// @leet end

File: leetcode/105.construct-binary-tree-from-preorder-and-inorder-traversal.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        in_idx_map = {k: i for i, k in enumerate(inorder)}

        def helper(
            pre_root_idx: int, in_left_idx: int, in_right_idx: int
        ) -> TreeNode | None:
            if in_left_idx > in_right_idx:
                return None

            val = preorder[pre_root_idx]
            in_root_idx = in_idx_map[val]
            k = in_root_idx - in_left_idx  # k = left tree size

            return TreeNode(
                val,
                helper(pre_root_idx + 1, in_left_idx, in_root_idx - 1),
                helper(pre_root_idx + k + 1, in_root_idx + 1, in_right_idx),
            )

        return helper(0, 0, len(inorder) - 1)


# @leet end
